doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,208 @@
+# Source Metrics for Users, v2
+
+- Associated Epic: https://github.com/MaterializeInc/materialize/issues/23154
+
+
+## The Problem
+
+The [existing source statistics](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_statistics_per_worker)
+we expose to users for sources answer _some important questions_, like:
+- _Generally_, how fast we are reading data from upstream?
+- Have we committed the source's snapshot?
+
+but fail to answer **two** critical questions users regularly ask about their sources:
+
+1. How much of its snapshot has this source processed?
+2. Is this source keeping up with _upstream_? (This is usually related to: can I resize this source?)
+
+
+This design document proposes metrics that answer these questions, and goes into detail about their
+implementation.
+
+## Success Criteria
+
+This design has succeeded when we are able to present these metrics to users in the console,
+and they can use them to answer questions about their source's snapshotting and sizing.
+
+## Out of Scope
+
+- Expose new metrics for sinks.
+- Solve the _hydration backpressure problem_ (more on this later)
+- Expose detailed, _granular_ metrics that are 100% accurate.
+
+## Solution Proposal
+
+### Frontier Aggregation
+
+Sources define progress as frontiers within a source-defined timestamp type. These times
+can be partially ordered, and various shapes. In the context of metrics, we want to
+be able to _calculate rates_ and _compare progress against a total_ in a way consistent
+across all source.
+
+This means sources will need to implement a way to aggregate their notions of progress into
+a single 64-bit unsigned integer. Clients processing this data can then consistently calculate rates
+and percentages from these numbers. Note that the _unit_ of the single integer does not need to be
+consistent across sources.
+
+#### Kafka frontier aggregation
+
+For example, consider the following kafka source frontiers:
+```
+# Partition Id: Offset
+f1: {0: 100, 1: 150}
+```
+
+A kafka source with this frontier can be said to have processed 240 offets.
+
+#### Postgres frontier aggregation
+
+The Postgres source has 2 notions of progress. During snapshotting it can sum the number
```

Comment:
This also is unclear as to what it means. The postgres source has only one notion of progress, a frontier of LSN